src/transformers/models/encodec/modeling_encodec.py

```diff
@@ -411,21 +413,31 @@ def get_num_quantizers_for_bandwidth(self, bandwidth: Optional[float] = None) ->
             num_quantizers = int(max(1, math.floor(bandwidth * 1000 / bw_per_q)))
         return num_quantizers
 
-    def encode(self, embeddings: torch.Tensor, bandwidth: Optional[float] = None) -> torch.Tensor:
+    def encode(self, embeddings: torch.Tensor, bandwidth: Optional[float] = None) -> Tuple[torch.Tensor, List]:
         """
         Encode a given input tensor with the specified frame rate at the given bandwidth. The RVQ encode method sets
         the appropriate number of quantizers to use and returns indices for each quantizer.
         """
         num_quantizers = self.get_num_quantizers_for_bandwidth(bandwidth)
         residual = embeddings
         all_indices = []
+        quantization_steps = []
         for layer in self.layers[:num_quantizers]:
             indices = layer.encode(residual)
             quantized = layer.decode(indices)
+
+            if self.training:
+                # Pass the gradients straight through the quantization, by directly linking the gradients of the input
+                # embed_ind with the output quantize in the computation graph.
+                quantized = residual + (quantized - residual).detach()
+                quantization_steps.append((residual, quantized.detach()))
+
+            # Note: There may be a bug here with the quantized results, but we do not fix it as it is present in the
+            # original FB code as well. For more context, see https://github.com/facebookresearch/encodec/issues/25.
             residual = residual - quantized
             all_indices.append(indices)
         out_indices = torch.stack(all_indices)
-        return out_indices
+        return out_indices, quantization_steps
```

Comment:
Can we convert `quantization_steps` to a tensor here? Like we do `all_indices`? Makes it easier to `torch.compile` a module!